torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -491,6 +495,154 @@ def codegen_define(self, kernel: CppTemplateKernel) -> str:
+    ),
+)
+class CppFP16MicroGemm(CppMicroGemmFP32Vec):
+    """
+    This class generates the code for micro gemm using FP16 vec instructions for compute.
+    Both compute & accum are in FP16.
+    A should be float16 and B should be int8.
+    Scale is fused because overflow would happen during accum if scale is a post-op.
+    """
+
+    TEMPLATE_ENTRY = r"""
+{{declare_kernel}} {
+    {{kernel.assert_function}}(N % {{block_n}} == 0, "N dimension must be multiple of {{block_n}}");
+    {{kernel.assert_function}}(K % {{block_k}} == 0, "K dimension must be multiple of {{block_k}}");
+    // TODO(jgong5): loop unroll for M and N
+    for (int64_t m = 0; m < M; m += {{block_m}}) {
+        int64_t block_m = std::min<int64_t>(M - m, {{block_m}});
+        for (int64_t n = 0; n < N; n += {{block_n}}) {
+            if (block_m == {{block_m}}) {
+                {{kernel_name}}_kernel<{{block_m}}, {{block_n}}, accum>(
+                    A + m * lda,
+                    B + n,
+                    S + n,
+                    C + m * ldc + n,
+                    K,
+                    lda,
+                    ldb,
+                    ldc
+                );
+            } else {
+                switch (block_m) {
+{%- for b in range(block_m - 1, 0, -1) %}
+                case {{b}}:
+                    {{kernel_name}}_kernel<{{b}}, {{block_n}}, accum>(
+                        A + m * lda,
+                        B + n,
+                        S + n,
+                        C + m * ldc + n,
+                        K,
+                        lda,
+                        ldb,
+                        ldc
+                    );
+                    break;
+{%- endfor %}
+                default:
+                    {{kernel.assert_function}}(false, "Unsupported block_m: {{block_m}}");
+                }
+            }
+        }
+    }
+}
+"""
+
+    TEMPLATE_KERNEL = r"""
+template <int64_t BLOCK_M, int64_t BLOCK_N, bool accum>
+inline void {{kernel_name}}_kernel(
+    const {{input_t}}* {{restrict_keyword}} A,
+    const {{input2_t}}* {{restrict_keyword}} B,
+    const {{input_t}}*{{restrict_keyword}} S,
+    {{output_t}}* {{restrict_keyword}} C,
+    int64_t K,
+    int64_t lda,
+    int64_t ldb,
+    int64_t ldc
+) {
+    constexpr auto VLEN = 32;
+    constexpr auto ROWS = BLOCK_M;
+    constexpr auto COLS = BLOCK_N / VLEN;
+
+    __m512h va;
+    __m512h vb[COLS];
+    __m512h scales[COLS];
+    __m512h vc[ROWS * COLS];
+
+    auto load_scales = [&](int i) {
+        scales[i] = _mm512_loadu_ph((__m512h*)(S + i * VLEN));
+    };
+    c10::ForcedUnroll<COLS>{}(load_scales);
+
+    auto loadc = [&](auto i) {
+        vc[i] = _mm512_set1_ph(0.0f);
+    };
+    c10::ForcedUnroll<ROWS * COLS>{}(loadc);
+
+    auto compute = [&, COLS](auto i, int k) {
+        constexpr int row = i / COLS;
+        constexpr int col = i % COLS;
+
+        if constexpr (col == 0) {
+            va = _mm512_set1_ph(*((_Float16*)(A + row * lda + k)));
```

Comment:
A very interesting observation is that changing `*((_Float16*)(A + row * lda + k))` to `A[ row * lda + k]` increases latency more than 2x for GEMMs in LLaMA2.

I haven't dug into the reason yet & can do that later, since it doesn't affect this PR, but I thought I might as well document this behavior, so that it's searchable.

Reply:
Hi @jgong5, do you know why using `(_Float16*)` makes so much performance difference? Thanks!

Reply:
Perhaps we can check the assembly code of these two versions to see if any extra instructions have been compiled.